File: sorting/insertion/insertion-sort.rs
fn main() 
{
    let mut a: Vec<i32> = vec![ -100, 2000, 5, 2, 7, 4, 15, 0 ];
    println!("Before sorting: {:?}", a);


    for j in 1..a.len() {
        let key: i32 = a[j];

        let mut i: usize = j - 1;
        let mut k: usize = i + 1;

        while k > 0 && a[i] > key {
            // Shift elements that are greater forward in the list
            a[k] = a[i];

            k = i;
            if i > 0 {
                i = i - 1;
            }
        }

        a[k] = key;
    }

    println!("After sorting: {:?}", a);

}

File: sorting/insertion/insertion-sort.c
#include<stdio.h>

/** Performs insertion sort on a[] and array of length 'len'
 */
void insertion_sort(int a[], int len)
{
  // sort through an array
  int j;
  for (j = 1; j < len; j = j + 1)
  {
    // current element we are moving
    int current = a[j];

    int i = j - 1;
    while (i >= 0 && a[i] > current) 
    {
      // Shift elements forward if they are greater than current element
      a[i+1] = a[i];
      i = i - 1;
    }

    // Put current element in the last spot we shifted from
    a[i + 1] = current;
  }

}

/** Helper that prints out an array
 */
void print_array(char msg[], int a[], int len) {

  int i;
  printf("%s", msg);
  for (i = 0; i < len; i = i + 1)
  {
    printf("%d ", a[i]);
  }

  printf("\n");
}

/** Create an array and sort it!
 */
int main() 
{
  int a[] = {1, 3, 1000, -20, 0};
  int len = sizeof(a)/sizeof(a[0]);

  print_array("Before sort: ", a, len);
  insertion_sort(a, len);
  print_array("After sort: ", a, len);

  return 0;
}


